Feature/cache expiration

gsantoro opened the pull request:
## What does this PR do?

Make the internal cache used by various kubernetes metricset to never expire so as to avoid missing metrics or calculate incorrect metrics as explained in this issue https://github.com/elastic/beats/issues/31778.

## Why is it important?

It fixes a bug reported by a customer at https://github.com/elastic/sdh-beats/issues/2098

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- [ ] Manually debug behaviour by using the instructions at https://github.com/elastic/beats/pull/31748

## Related issues

Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes [#123](https://github.com/elastic/beats/pull/31748)
- Relates https://github.com/elastic/sdh-beats/issues/2098

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31785/detail/PR-31785/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31785/detail/PR-31785/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31785/detail/PR-31785/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31785/detail/PR-31785/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_31785.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-05-31T15:00:37.668Z&rangeTo=2022-05-31T15:20:37.668Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=95ee760903e42b010eb0cc74504ae300&transactionId=f88e9f494dbba955)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-05-31T15:10:37.668+0000


* Duration: 16 min 21 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31785/detail/PR-31785/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `deploy/kubernetes-checks - make -C deploy/kubernetes all;make check-no-changes;`
<ul>
<li>Took 0 min 4 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31785/runs/1/steps/1480/log/?start=0">here</a></li>
<li>Description: <code>make -C deploy/kubernetes all;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `metricbeat-checks - make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat c`
<ul>
<li>Took 3 min 51 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31785/runs/1/steps/1474/log/?start=0">here</a></li>
<li>Description: <code>make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-checks - make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -`
<ul>
<li>Took 3 min 54 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31785/runs/1/steps/1199/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -C metricbeat check;make -C metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31785/runs/1/steps/1491/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

gizas commented on deploy/kubernetes/metricbeat-kubernetes.yaml:
```diff
@@ -83,7 +84,7 @@ metadata:
   system.yml: |-
     - module: system
-      period: 10s
+      period: 1m
```
Please revert those values to original ones before merge

ChrsMark reviewed (commented):
A couple of comments we already discussed offline:

1. Can you revert the changes on the config files that are not needed for the patch itself?
2. In a quick view, I think we can use your logic to set the expiration’s default value and make it also available as an optional setting. In this, our users can just ignore this setting and the defaults will do the trick. If in any case someone wants to specify this setting there will be an option available.

Some key code parts that might help in adding the setting:

1. https://github.com/elastic/beats/blob/9535a581641cedddff20ba22dfa092383d3a7e33/metricbeat/module/kubernetes/util/kubernetes.go#L52
2.  most probably you will need to unpack the config at the metricset level like what we do inside the `Enricher`'s implementation: https://github.com/elastic/beats/blob/9535a581641cedddff20ba22dfa092383d3a7e33/metricbeat/module/kubernetes/util/kubernetes.go#L186
3. Check `Enricher`'s initialisation: https://github.com/elastic/beats/blob/89bcc33a9a90cbacceaa61fceec0e8d23073574c/metricbeat/module/kubernetes/container/container.go#L78

gsantoro commented:
I put `scraping.period * 2` as an initial placeholder. I wasn't sure about 2 as a final value. Maybe we should have `scraping.period * 5` to be more resilient. I don't have a strong opinion on the multiplier value.
My point is though, that this behaviour should be the same no matter the length of the scraping period. The value in the cache is replaced at every successful scrape and the expiration updated to a full timeout (cache period), even if the entries are not expired, as long as there has been at least 1 full period from last scraping or it is the first scraping

> if statsCache.lastFetchTimestamp.IsZero() || now.Sub(statsCache.lastFetchTimestamp) > m.Config().Period { ... http.FetchContent() ... }

A cache expiration time of 5 * scraping.Period also prevent that the cache expires if one or more scraping do NOT take place. 

scenarios with current code (before modifications in this PR):
- Example 1
  - scrape every 10s
  - cache expiration at 2m
  - cache is replaced every 10s (at every scrape)
  - cache expires 2m in the future of the last scrape
  - metrics are correctly calculated since the cache is always full

- Example 2 (failure discovered in SDH)
  - scrape every 10m
  - cache expiration at 2m
  - cache expires 2m after the last scrape. it stays empty for 8m, metricbeat read an empty cache and it calculates wrong metrics
  - cache is refilled but only read at the next period after 10m when it is already empty again.

Scenarios with modifications in this PR:
- Example 1
  - scrape at 10s
  - cache expiration at 50s
  - with no scrape failures, cache is replenished every 10s and it expires 50s in the future of the last scrape
  - if some scrapes fails to run, if there is at least 1 scrape every 50s, the cache expires 50s in the future of the last scrape and the right metrics are calculates. You can have up to 4/5 failed scrapes and still behave correctly

- Example 2
  - scrape at 10m
  - cache expiration at 50m
  - similarly to example 1. Cache never expires, it is resilient to scraping failures
  - the user doesn't need to guess what it a good value for the cache expiration

There is a similar config for the expiration of the cache at `beats/x-pack/metricbeat/module/prometheus/collector/data.go` at line 29 `counters := collector.NewCounterCache(base.Module().Config().Period * 5)`

gsantoro commented:
Comments on https://github.com/elastic/beats/pull/31785#pullrequestreview-991637896
- I will revert the changes. I left them there for you guys to check the behaviour of the new code
- I agree we should expose an optional config for the user to customise the behaviour of the cache and do the trick with `cache_expiration = period * 5` otherwise. Thanks for the suggestions. I'll have a look

ChrsMark commented:
> Comments on [#31785 (review)](https://github.com/elastic/beats/pull/31785#pullrequestreview-991637896)
> 
> * I will revert the changes. I left them there for you guys to check the behaviour of the new code

In that case you can just add everything needed at `How to test this PR locally` section in the PR's description (see for example https://github.com/elastic/beats/pull/25640).

ChrsMark commented on metricbeat/module/kubernetes/util/kubernetes.go:
```diff
@@ -329,21 +339,67 @@ func GetDefaultDisabledMetaConfig() *kubernetesConfig {
+	if config.CacheExpirationTime <= 0 {
+		if cacheTimeout < minCacheExpirationTime {
+			config.CacheExpirationTime = minCacheExpirationTime
+			logp.Debug("setting CacheExpirationTime = minCacheExpirationTime. CacheExpirationTime: %s", config.CacheExpirationTime.String())
```
I would use sth more user friendly like `setting cache_expiration_time to minimum default: %s`

ChrsMark commented on metricbeat/module/kubernetes/util/kubernetes.go:
```diff
@@ -329,21 +339,67 @@ func GetDefaultDisabledMetaConfig() *kubernetesConfig {
+
+		} else {
+			config.CacheExpirationTime = cacheTimeout
+			logp.Debug("setting CacheExpirationTime = DefaultCacheTimeout. CacheExpirationTime: %s", config.CacheExpirationTime.String())
```
Similar comment here about the wording.

ChrsMark commented on metricbeat/module/kubernetes/util/kubernetes.go:
```diff
@@ -53,8 +54,9 @@ type kubernetesConfig struct {
-	SyncPeriod time.Duration `config:"sync_period"`
+	Node                string        `config:"node"`
+	SyncPeriod          time.Duration `config:"sync_period"`
+	CacheExpirationTime time.Duration `config:"cache_expiration_time"`
```
This would also need to be added and explained in the respective docs.

MichaelKatsoulis replied:
https://github.com/elastic/beats/blob/d983d4a00aef40fd5fac4188dd4444a89da7b282/metricbeat/docs/modules/kubernetes.asciidoc

ChrsMark commented on metricbeat/module/kubernetes/util/kubernetes.go:
```diff
@@ -329,21 +339,67 @@ func GetDefaultDisabledMetaConfig() *kubernetesConfig {
+func validateConfig(config *kubernetesConfig, moduleConfig mb.ModuleConfig, cacheTimeout time.Duration) (*kubernetesConfig, error) {
+	// NOTE: minimum cache expiration time > period avoid that the cache expires just
+	// before next scraping time. Period * 2 is arbitrary but it a good buffer to avoid
+	// that a late scraping run would affect the metrics (for the absence of entries in the cache)
```
Some typos or misspellings here I guess. Also consider adding a link to the public GH issue that originally reported this edge case.

gsantoro commented on deploy/kubernetes/metricbeat-kubernetes.yaml:
```diff
@@ -156,7 +156,7 @@ spec:
       containers:
       - name: metricbeat
-        image: docker.elastic.co/beats/metricbeat:8.4.0
+        image: docker.elastic.co/beats/metricbeat:8.3.0-SNAPSHOT
```
my bad, it should be gone now

ChrsMark commented on metricbeat/module/kubernetes/util/kubernetes.go:
```diff
@@ -82,12 +84,16 @@ func NewResourceMetadataEnricher(
 	}
 
+	if PerfMetrics.Timeout < config.CacheExpirationTime {
+		PerfMetrics = NewPerfMetricsCache(config.CacheExpirationTime)
```
Will this overwrite the values if `NewResourceMetadataEnricher` is called more than once? Like at [state_container](https://github.com/elastic/beats/blob/89bcc33a9a90cbacceaa61fceec0e8d23073574c/metricbeat/module/kubernetes/state_container/state_container.go#L121) and [container](https://github.com/elastic/beats/blob/89bcc33a9a90cbacceaa61fceec0e8d23073574c/metricbeat/module/kubernetes/container/container.go#L78) ?

The result should be the same but you just initialise the cache twice and the second call to `NewContainerMetadataEnricher` overwrites the already existent value/object/reference.

A workaround here would be to initialise the cache globally like it was happening already and then use `SetExpirationTime()` method to set the expiration. This new method can be called multiple times but needs to ensure that only updates the timeout if the incoming value is bigger than the already set. In this was you ensure that you use the `max()` of the various scrape.periods.

gsantoro commented:
Hey guys, 
I have applied the latest changes:
- cache is created at module level once with 0 timeout
- the final cache timeout = max(Period) * 2
- if you find a longer period in one of the metricset configs, the timeout in the cache is changed in place (without recreating the cache object), and the janitors are restarted
- I added setter and getter to change the cache timeout
- fixed a bug when StopJanitor

ChrsMark commented on metricbeat/module/kubernetes/util/metrics_cache.go:
```diff
@@ -68,6 +59,41 @@ func (c *PerfMetricsCache) Stop() {
 	c.ContainerCoresLimit.Stop()
 }
 
+func (c *PerfMetricsCache) GetTimeout() time.Duration {
```
Could you add a docstring here to document this method and the logic encapsulated?

ChrsMark commented on metricbeat/module/kubernetes/util/metrics_cache.go:
```diff
@@ -68,6 +59,41 @@ func (c *PerfMetricsCache) Stop() {
+	return ans
+}
+
+func (c *PerfMetricsCache) SetTimeout(timeout time.Duration) {
```
nit: 
```suggestion
func (c *PerfMetricsCache) SetOrUpdateTimeout(timeout time.Duration) {
```

Also could you please add a docstring for this method?

ChrsMark reviewed (approved):
👌🏼 Very nice! I left 2 minor comments but overall you can consider it ready!

ChrsMark commented:
A changelog entry is also needed for this.

tetianakravchenko commented:
I think other backport labels for other versions (v8.0.0-v8.3.0) should be added here

ChrsMark reviewed (changes requested):
Make sure comments are addressed with commits before merging specifically for the changelog.

ChrsMark commented on CHANGELOG.next.asciidoc:
```diff
@@ -57,6 +57,7 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 *Metricbeat*
 
 - make `system/filesystem` code sensitive to `hostfs` and migrate libraries to `elastic-agent-opts` {pull}31001[31001]
+- make the internal cache PerfMetricsCache in the kubernetes module to never expire before scraping. This avoid metrics like `kubernetes.container.cpu.usage.limit.pct` from not being populated.
```
```suggestion
- Fix kubernetes module's internal cache expiration issue. This avoid metrics like `kubernetes.container.cpu.usage.limit.pct` from not being populated. {pull}31785[31785]
```

ChrsMark reviewed (changes requested):
A minor suggestion/fix for the changelog. Keep in mind that changelog entries are user facing content and thus should be quite accurate and not super technical.

ChrsMark reviewed (approved):
Thanks for the detailed work here!
Wait for green CI and ship it! :shipit:

gizas reviewed (approved):
Only missing a screenshot after the implementation to see that final metrics actually appear.

Also is there a simple way to verify the value of cache timeout without debugging?

gsantoro commented on deploy/kubernetes/metricbeat-kubernetes.yaml:
```diff
@@ -156,12 +157,18 @@ spec:
-        image: docker.elastic.co/beats/metricbeat:8.4.0
+        # image: docker.elastic.co/beats/metricbeat:8.3.0-SNAPSHOT
+        image: metricbeat-debugger-image:latest
+        imagePullPolicy: Never
```
you are looking at an outdated commit. this has been resolved already

gsantoro commented on deploy/kubernetes/metricbeat-kubernetes.yaml:
```diff
@@ -27,7 +27,7 @@ data:
                 - module: kubernetes
                   hosts: ["kube-state-metrics:8080"]
-                  period: 10s
+                  period: 3m
```
you are looking at an outdated commit. this has been resolved already
